RFC: Speedier, simpler and more systematic index conversions

mbauman opened the pull request:
This patch dramatically simplifies the fallback indexing algorithms for arrays
by standardizing the way in which the passed indices are converted `to_index`
and by rigorously defining what a supported index type is.

The very first thing that occurs within the fallback indexing methods is that
the passed indices are converted to a supported indexing type by
`to_indices(...)`. A supported index type is either `Int` or an `AbstractArray`
of indices. This means that it is at this point that both `Colon` and arrays of
booleans get converted to collections of indices (as specialized array types:
`Slice` and `LogicalIndex`, respectively). This *drastically* simplifies much
of the internal logic. Whereas we had previously encoded the special behaviors
of `:` indexing in about four places, this patch does it once and then stores
the result in a `Slice` AbstractVector. Similarly, `CartesianIndex` is now
converted within `to_indices`, as well.

In addition to simplifying the internal definitions, this patch also makes it
easier for both external packages and Base to add new behaviors in a systematic
fashion. Simple extensions would include things like `Not` indices and
[EndpointRanges.jl](https://github.com/JuliaArrays/EndpointRanges.jl).

Additionally, logical indexing is now represented by a specialized
`LogicalIndex` type that enables fast iteration over the boolean mask without
allocations. This allows index conversions to occur before checking bounds,
and should enable further optimizations for LinearSlow arrays.

 ### Breaking Behaviors:

* The stored type of indices in SubArrays change from `Colon` to `Slice`.

### Non-breaking behavior changes:

* Fully deprecates using `Colon` as pseudo-collections of integers.
* Logical indices are converted to a lazy, iterative collection of indices that doesn't support indexing.  A deprecation provides indexing with O(n) search.

### TODO:

* [x] CI pass
* [x] @nanosoldier `runbenchmarks(ALL, vs = ":master")`
* [x] Decide: should we export and formally document `to_indices`? **YES**. I think I'd be inclined to continue to *not* export `to_index`, particularly now that it's not the entire story.  You *must* call `to_indices` to get the complete conversion, but you can extend `to_index` to provide simpler customizations. **Exported to_indices**.

mbauman committed:
Simpler and more systematic index conversions

This patch dramatically simplifies the fallback indexing algorithms for arrays
by standardizing the way in which the passed indices are converted `to_index`
and by rigorously defining what a supported index type is.

The very first thing that occurs within the fallback indexing methods is that
the passed indices are converted to a supported indexing type by
`to_indices(...)`. A supported index type is either `Int` or an `AbstractArray`
of indices. This means that it is at this point that both `Colon` and arrays of
booleans get converted to collections of indices (as specialized array types:
`Slice` and `LogicalIndex`, respectively). This *drastically* simplifies much
of the internal logic. Whereas we had previously encoded the special behaviors
of `:` indexing in about four places, this patch does it once and then stores
the result in a `Slice` AbstractVector. Similarly, `CartesianIndex` is now
converted within `to_indices`, as well.

In addition to simplifying the internal definitions, this patch also makes it
easier for both external packages and Base to add new behaviors in a systematic
fashion. Simple extensions would include things like `Not` indices and
[EndpointRanges.jl](https://github.com/JuliaArrays/EndpointRanges.jl).

Additionally, logical indexing is now represented by a specialized
`LogicalIndex` type that enables fast iteration over the boolean mask without
allocations. This allows index conversions to occur before checking bounds,
and should enable further optimizations for LinearSlow arrays.

 ### Breaking Behaviors:

* The stored type of indices in SubArrays change from `Colon` to `Slice`.

mbauman commented:
I'd just like to note that this PR is at a net-negative LOC… and that includes ~30 lines of deprecations and ~40 lines of docstrings. :)

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19730) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/9aaffc8bf42c1c893b2917b9a7fed1b3ea5ea8e3/85beb4b_vs_915a07d/report.md). cc @jrevels

mbauman committed:
Attempt to address performance regressions

mbauman commented:
Let's see if Nanosoldier is any happier with the latest iteration. @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19730#issuecomment-269565065) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6735ed11a95e79a4267f258cabe390bc04776ab5/f184104_vs_309657f/report.md). cc @jrevels

mbauman committed:
More perf improvements

* Move non-inlined function boundary up one method in the nonscalar indexing call chain and simplify the setindex call chain.
* Avoid calling indices(A) when it is known to be not needed. It'd be nice if Julia could identify this, but this is a very simple workaround.

mbauman commented:
Third time's the charm?  @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19730#issuecomment-269662049) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3297c17fd747b83886534f19e1c316fd0160f05c/131417d_vs_b561cfb/report.md). cc @jrevels

mbauman committed:
Type stable concatenation

Because the abstract fallbacks now rely more upon dispatch, the effect of a type-instability in the indices is greater.

mbauman committed:
Specialize indexing into ::Array with mixtures of Integers and CartesianIndices.

This is a little unfortunate that we no longer try to call the canonical method multiple times when indexing with CartesianIndex, but I think the standardization is worth it.

mbauman commented:
Alright, this is looking pretty good.  I think I should have mitigated the remaining regressions.  One unfortunate thing is that this increases the impact of a type-instability within indexing code — hence the cat regressions.

andreasnoack commented:
So the `broadcast`+`fusion` regression is gone with the latest changes?

mbauman commented:
Yes, I believe so. I'm feeling a little bad about co-opting nanosoldier time, but we should run it again. I can't reproduce the fusion regression locally on my old hardware so it's hard for me to debug it.

I should sell the case for the refactor a little better, but I strongly believe it's a much better and saner design.

tkelman commented:
> I'm feeling a little bad about co-opting nanosoldier time

Please don't, this is exactly what it's for and why we're going a little out of our way to try not to break it right now. If there are a lot of people using it, it queues jobs, which is fine. I don't think the nanosoldier queue has ever been especially deep.

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

jrevels commented:
retriggering Nanosoldier, ref [my comment here](https://github.com/JuliaLang/julia/pull/19745#issuecomment-269774174)

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/19730#issuecomment-269775294):
```
NanosoldierError: failed to run benchmarks against primary commit: ErrorException("failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(139)) [139]")
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5c34fa725bc7da5cae9af8c563c6baf3db33f443/04b22eb_vs_6ad62c4/report.md)
cc @jrevels

jrevels commented:
Looks like Nanosoldier's benchmark script [segfaulted](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5c34fa725bc7da5cae9af8c563c6baf3db33f443/04b22eb_vs_6ad62c4/logs/04b22eb0c5233add842f424bb1a86371abc34ab0_primary.err). The script it generates can be found [here](https://github.com/JuliaCI/Nanosoldier.jl/blob/master/src/jobs/BenchmarkJob.jl#L318), we can see from the [STDOUT log](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5c34fa725bc7da5cae9af8c563c6baf3db33f443/04b22eb_vs_6ad62c4/logs/04b22eb0c5233add842f424bb1a86371abc34ab0_primary.out) that it at least executed the `cset` command before it died.

mbauman committed:
Fix and test views with logical masks

tkelman commented on base/abstractarray.jl:
```diff
@@ -873,49 +875,44 @@ end
     r
 end
-function _setindex!(::LinearFast, A::AbstractVector, v, I1::Real, I::Real...)
+function _setindex!(::LinearFast, A::AbstractVector, v, I1::Int, I::Int...)
```
Am I right that this signature is for allowing extra trailing 1's?

mbauman replied:
Yup.

tkelman commented on base/abstractarray.jl:
```diff
@@ -1143,23 +1140,29 @@ function cat_t(dims, T::Type, X...)
     return _cat(A, shape, catdims, X...)
 end
 
+_cat_inds(i, x, ::Tuple{}, offsets, concat) = ()
```
could do with some comments, what these recursions are going for?

mbauman replied:
This change is more than is necessary here… and probably has compile-time performance implications.  I reverted it in favor of a simpler work-around.

tkelman commented on base/operators.jl:
```diff
@@ -818,19 +818,83 @@ function setindex_shape_check{T}(X::AbstractArray{T,2}, i::Integer, j::Integer)
+"""
+    to_index(A, i)
+
+Convert index `i` for use into indexing into array `A`. Custom array types
```
"for use in indexing" ?

tkelman commented on base/operators.jl:
```diff
@@ -818,19 +818,83 @@ function setindex_shape_check{T}(X::AbstractArray{T,2}, i::Integer, j::Integer)
+index `i`. More complicated index types, however, require more context about
+the dimension into which they index. To support those cases, `to_indices(A, I)`
+calls `to_indices(A, indices(A), I)`, which then recursively walks through both
+the given tuple of indices and the indices in `A` in tandem.
```
"indices in `A`" or "indices of `A`" ?

tkelman commented:
This is pretty neat. Hasn't gotten a whole lot of review, but should we be trying to get it in right now as we're finalizing feature freeze for 0.6?

mbauman committed:
Revert "Type stable concatenation"

This reverts commit f15ad8817cff869be6973f051a6c63faf5fd6b39.

mbauman committed:
Simpler workaround for concatenation type-stability

mbauman commented:
Thanks, Tony. I think the standardization here is really valuable. It'd be nice if this could make the cut.

That said, I'd like some feedback on the API of `to_indices` and `to_index`. This PR makes `to_indices` great to use as a caller.  Note how it will automatically make `view` work with any custom index type implements its conversion within `to_indices`.  This will make it *much* easier to add custom index types and have them just work everywhere.  It's also really nice internally; we only need to worry about indices of type `Int` and `AbstractArray` within the all internal functions.

The trouble is that implementing a custom `to_indices` specialization is rather messy because it has to fit into a complicated recursion setup.  For example, here's a simplistic take on `Not` indexing:

```jl
julia> immutable Not{T}
           vals::T
       end

julia> import Base: to_indices, uncolon, tail, _maybetail
       @inline to_indices(A, inds, I::Tuple{Not, Vararg{Any}}) =
           (setdiff(uncolon(inds, (:, tail(I)...)), I[1].vals), to_indices(A, _maybetail(inds), tail(I))...)
to_indices (generic function with 9 methods)

julia> A = reshape(1:12, 3, 4);
       A[Not(2), Not(2:3)]
2×2 Array{Int64,2}:
 1  10
 3  12

julia> view(A, Not(2), Not(2:3))
2×2 SubArray{Int64,2,Base.ReshapedArray{Int64,2,UnitRange{Int64},Tuple{}},Tuple{Array{Int64,1},Array{Int64,1}},false}:
 1  10
 3  12
```

Of course, some of this is nasty because of generalized linear indexing; that's why I'm leaning on the internal `uncolon`. But the recursion alone is tricky.  A simpler example that's still tough to get right is if a custom array wishes to support scalar indexing with non-integers.  For example, to support the `SymInt` example I posited in #12567, you can "just" protect the symbols within a 0-dimensional array:

```jl
import Base: to_indices, _maybetail, tail
@inline to_indices(A::SymInt, inds, I::Tuple{Symbol, Vararg{Any}}) =
    (fill(I[1]), to_indices(A, _maybetail(inds), tail(I))…)
```

So to make this simpler, I added fallbacks to the existing `to_index` function, allowing you to not need to worry about recursion or the parent's indices if you don't need to.  So for the symbol example, you can simply write `to_index(::SymInt, s::Symbol) = fill(s)`.  I don't really like this, though, since *not all indices* will fall back here.  Only those indices that don't need the extra information will hit `to_index`.  I could see someone getting really frustrated when their custom `to_index(::MyArray, ::Colon)` isn't getting called.  Or a more 

So we can either document that extending `to_index` might be crazy-making if someone has intercepted your method within `to_indices`… or we can nix `to_index` entirely and just document how to extend `to_indices` very carefully.  Thoughts?

mbauman committed:
Minor docstring fixups from review

timholy commented on base/multidimensional.jl:
```diff
@@ -216,70 +219,100 @@ end
+    mask::A
+    sum::Int
+    LogicalIndex(mask::A) = new(mask, countnz(mask))
+end
```
Will be interesting to see whether there are any performance consequences---could potentially be an extra trip through `mask` for some algorithms.

mbauman replied:
Yes; in general I was surprised how hard I had to fight to match the allocating `find` algorithm.  `find(::BitArray)` is wonderfully optimized and tough to beat.

timholy commented on base/operators.jl:
```diff
@@ -818,19 +818,83 @@ function setindex_shape_check{T}(X::AbstractArray{T,2}, i::Integer, j::Integer)
+calls `to_indices(A, indices(A), I)`, which then recursively walks through both
+the given tuple of indices and the dimensional indices of `A` in tandem.
+"""
+to_indices(A, I::Tuple) = (@_inline_meta; to_indices(A, indices(A), I))
```
Do you really need to pass `A` once you've called `indices(A)`?

mbauman replied:
No, but the advantage is that it allows for a simpler `to_index(A, i)` method for array types to override.

timholy commented on base/multidimensional.jl:
```diff
@@ -288,84 +321,39 @@ getindex(t::Tuple, I...) = getindex(t, IteratorsMD.flatten(I)...)
-@inline function _getindex{N}(l::LinearIndexing, A::AbstractArray, I::Vararg{Union{Real, AbstractArray, Colon},N}) # TODO: DEPRECATE FOR #14770
-    @boundscheck checkbounds(A, I...)
-    _unsafe_getindex(l, _maybe_reshape(l, A, I), I...)
+@generated function _getindex(l::LinearIndexing, A::AbstractArray, I::Union{Real, AbstractArray}...)
```
`Real` or `Int`?

mbauman replied:
Yes, this could be tightened eventually.  It's not strictly necessary, though.

timholy reviewed (approved):
I like it. A couple of questions:
- I'm a little confused about transforming Colon->Slice, can you explain a bit why that's necessary?
- You haven't by any chance tested the impact on AxisArrays.jl and Interpolations.jl, have you?

mbauman commented:
Thanks for looking at this, Tim!  With regards to your two primary comments:

> * I'm a little confused about transforming Colon->Slice, can you explain a bit why that's necessary?

It's not strictly necessary, but there are a few reasons to do this:
1. It means that we no longer need to worry about Colon's special cases and context-sensitive behavior in multiple places.  Before, we had to worry about trailing Colons in three places: `index_lengths`, `index_shape`, `decolon`.  By converting them to an `AbstractUnitRange`, we can specify their behavior in one place and be done.
2. It means that *any* algorithm that codes for `AbstractArray` will be correct with respect to `Colon`.  Any specializations are now simply optimizations.  They're no longer necessary for correctness.  Before, Colons were pseudo-vectors.  Now that kinda/sorta behavior is deprecated, and we just convert them to a real vector.
3. It simplifies what we consider a valid "index type" by 33%.
4. It's a nice test-case to ensure that `to_indices` performs well and has enough context to do relatively complicated things.

> * You haven't by any chance tested the impact on AxisArrays.jl and Interpolations.jl, have you?

I have not, but I've had their use-cases in mind.  To be honest, though, I don't think this alone will be sufficient to improve (or break) either all that much.  Both are tricky in their own ways.  While AxisArrays could specialize and use `to_indices(::AxisArray, I)` to sort its indices, it will still want to define its own non-scalar indexing and views in order to smartly create the new axes for the returned object. Interpolations will have difficulty with the return type.  While it could easily override `to_index(::Interpolation, n::Number)  = fill(n)` to allow any Number through Base's nonscalar indexing implementation, it'll allocate memory using the `eltype` of the iterator without consideration of the indices.

In a sense, both of these issues are failings of `similar`.  Future work could introduce a new function `indexed_similar(A, I...) = similar(A, index_shape(I...))` (or maybe even allow passing the indices themselves directly to `similar`).  This would allow Interpolations to use promotion to get the return type correct and it'd allow AxisArrays to pre-allocate an AxisArray with the proper axes.

timholy commented:
`similar` now allows you to pass the indices. There's also `similar(dims->f(dims), inds)` which allows for a lot of flexibility in `similar`.

mbauman commented:
Isn't that a different sort of indices? Those define the offset-ness of each dimension, using `AbstractUnitRange`s to represent the indices of the indices.  I'm talking about passing the index tuple used by `getindex` directly… which would of course conflict with the current meaning.

Too many indices.

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

jrevels commented:
Had to kick nanosoldier, retriggering:

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19730#issuecomment-270123557) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d1694fc0b1e50ce52f405839412437eb53f63e25/488916d_vs_13b2254/report.md). cc @jrevels

tkelman commented:
bump for rebase (the conflicts are a bit nontrivial, would want you to double-check them), should we reassess whether there are any real regressions here worth worrying about?

mbauman commented:
Let's see how the performance looks atop my SubArray optimizations. There are definitely slight time regressions with linear indexing, but they're not terrible... and I think we should be able to optimize it further by going deeper into BitArray's internals.

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19730#issuecomment-270692749) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/db71eb1ca814fed455d16bb87e19992e8420e097/3abeffa_vs_1f59fbf/report.md). cc @jrevels

mbauman committed:
Further optimize iteration over LogicalIndex

mbauman committed:
Better deprecation messages

mbauman commented:
Can we go faster?  @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19730#issuecomment-270856424) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b79518d43c4d24b7ad2a3bef4cd2e621d6f64d66/a402bf1_vs_5b6ff4a/report.md). cc @jrevels

KristofferC commented:
Apparently :D

pabloferz commented on base/abstractarray.jl:
```diff
@@ -1173,7 +1170,7 @@ function cat_t(dims, T::Type, X...)
 
-function _cat(A, shape, catdims, X...)
+function _cat{N}(A, shape::NTuple{N}, catdims, X...)
     N = length(shape)
```
Remove this line?

mbauman replied:
Thanks!  Looks like it was throwing a warning that I missed.

mbauman committed:
Export to_indices and expand its documentation

mbauman committed:
Remove local variable N in cat

redundant with typevar and throwing a warning.

mbauman commented:
Alright, this is good to go as far as I'm concerned — It can be squashed upon merge.  I definitely think it's worth doing for 0.6.

vchuravy commented:
good to merge?

tkelman commented on base/essentials.jl:
```diff
@@ -220,6 +220,13 @@ function isassigned(v::SimpleVector, i::Int)
+Colons (:) are used to signify indexing entire objects or dimensions at once.
+Very few operations are defined on Colons directly; instead they are converted
+to `Slice`s upon indexing (within `to_indices`).
+"""
```
this should go into the stdlib doc index

tkelman commented:
ref https://github.com/JuliaLang/julia/pull/19730#discussion_r95390854

mbauman committed:
More careful consideration of multidimensional logical indices

mbauman committed:
Add Colon to stdlib; touch up its doc

vchuravy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19730#issuecomment-272016885) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/42d95477b4672f34d3e3055997adf45a816c87ed/aaea7ce_vs_93ecb41/report.md). cc @jrevels

oscardssmith commented:
Is it concerning that some of these tests took 0.00 times as long on this branch as master? That suggests to me, that something is broken. Assuming these were not a mistake, nanosoldier seems to quite like this.

mbauman commented:
The results here are definitely real.  This refactor allows Julia+LLVM to drastically reduce and/or eliminate the creation of some heap-allocated temporaries inside the loops.  This leads directly to improvements of several orders of magnitude since the only other "work" being done each iteration is a simple addition.  And with ranges it's even better than 100x: the loop becomes simple enough that it can be entirely replaced with a constant-time algorithm.

I've added two commits: one addresses your comment @tkelman, and the other adds an optimization for indexing into linear fast arrays with multidimensional logical masks by allowing the LogicalIndex to be a collection of linear indices in that case.

StefanKarpinski commented:
This is lovely work, @mbauman.

JeffBezanson commented:
Yes, this is an awesome improvement.

I still don't really understand the purpose of the `Slice` type. It seems very similar to `OneTo`. Could we normalize colons to those instead?

mbauman commented:
Thanks!

@JeffBezanson - Slice is really only required to support the current semantics of OffsetArrays, where indexing `A[:]` returns a similarly offset array. We use the offset-ness of the indices to determine the shape of the result, so in those cases `:` represents an offset vector. I suppose we could ask offset array packages to implement this, but I wanted to keep this non-breaking.

tkelman commented:
This has been causing an abstractarray test failure, but the exit status gets lost when tests run in parallel

mbauman commented:
That's a simple fix, but do we have an issue for `make test` missing failures?

StefanKarpinski commented:
> This has been causing an abstractarray test failure, but the exit status gets lost when tests run in parallel

Oops – that's two bugs. The lost exit status one is actually worse.

tkelman commented:
not yet, but it's been happening for at least a couple weeks. first thing I intend to try is disabling the anonymous module change Yichao made

Sacha0 commented:
@mbauman, this work seems worth a news entry. I'm afraid I wouldn't do it justice. Would you have time to sketch an appropriate entry? Best!
